File: binance4j-websocket/src/main/java/com/binance4j/websocket/client/WebsocketInterceptorCallback.java
package com.binance4j.websocket.client;

import com.binance4j.core.exception.ApiException;
import com.binance4j.websocket.callback.WebsocketCallback;
import com.binance4j.websocket.callback.WebsocketCloseObject;
import com.binance4j.websocket.event.WebsocketConnectionHandler;
import com.binance4j.websocket.event.WebsocketDisconnectionHandler;
import com.binance4j.websocket.event.WebsocketEventHandler;
import com.binance4j.websocket.event.WebsocketNoResponseHandler;

import lombok.AccessLevel;
import lombok.Data;
import lombok.Getter;
import okhttp3.Response;
import okhttp3.WebSocket;

/**
 * Interceptor callback for the {@link WebsocketClient} main callback to handle
 * event failures
 */
@Data
public class WebsocketInterceptorCallback<T> implements WebsocketCallback<T> {
	/** The main ws client. */
	WebsocketClient websocketClient;
	/** Event to handle connection failure and try to reconnect. */
	@Getter(value = AccessLevel.PROTECTED)
	private WebsocketEventHandler connectionHandler;
	/** Event to handle no response from server. */
	@Getter(value = AccessLevel.PROTECTED)
	private WebsocketEventHandler noResponseHandler;
	/** Event to handle failure in the disconnection state. */
	@Getter(value = AccessLevel.PROTECTED)
	private WebsocketEventHandler disconnectionHandler;

	/** Forces the call of onClosing and onClosed. */
	@Getter(value = AccessLevel.PROTECTED)
	private WebsocketEventHandler forceClosingHandler;

	/** Tells the interceptor if th e closing has been made by the client. */
	private boolean closedByClient;

	/** Has the onClosing handler been called. */
	private boolean onClosingCalled;

	@Getter(AccessLevel.NONE)
	private WebSocket socket;

	protected WebsocketCallback<T> callback;

	/**
	 *
	 * 
	 * @param websocketClient The main ws client
	 */
	public WebsocketInterceptorCallback(WebsocketClient websocketClient, WebsocketCallback<T> callback) {
		this.websocketClient = websocketClient;
		connectionHandler = new WebsocketConnectionHandler(websocketClient, this);
		noResponseHandler = new WebsocketNoResponseHandler(websocketClient, this);
		disconnectionHandler = new WebsocketDisconnectionHandler(websocketClient, this);
		this.callback = callback;
	}

	@Override
	public void onMessage(T payload) {
		noResponseHandler.run();
		callback.onMessage(payload);
	}

	@Override
	public void onClosing(WebsocketCloseObject closeObject) {
		if (forceClosingHandler != null) {
			forceClosingHandler.cancel();
		}
		onClosingCalled = true;
		// we force disconnection
		disconnectionHandler.run();
		callback.onClosing(closeObject);
	}

	@Override
	public void onClosed(WebsocketCloseObject closeObject) {
		// we stop disconnection forcing
		disconnectionHandler.cancel();
		callback.onClosed(closeObject);

		// Reconnect if server closed stream and keepAlive is on
		if (websocketClient.getConfiguration().isKeepAlive() && !closedByClient) {
			websocketClient.open();
		}

		socket = null;
	}

	@Override
	public void onOpen(Response response) {
		// we stop trying to reconnect
		connectionHandler.cancel();
		// we handle no response data
		noResponseHandler.run();

		callback.onOpen(response);
	}

	@Override
	public void onFailure(ApiException exception) {
		callback.onFailure(exception);
	}
}

File: connectors/src/main/kotlin/com/binance4j/websocket/callback/WebsocketInterceptorCallback.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.binance4j.websocket.callback

import com.binance4j.core.exception.ApiException
import com.binance4j.websocket.client.WebsocketClient
import com.binance4j.websocket.event.WebsocketConnectionHandler
import com.binance4j.websocket.event.WebsocketDisconnectionHandler
import com.binance4j.websocket.event.WebsocketEventHandler
import com.binance4j.websocket.event.WebsocketNoResponseHandler
import okhttp3.Response

import okhttp3.WebSocket

/**
 * Interceptor callback for {@link WebsocketClient} main callback to handle
 * event failures
 * @property websocketClient    The main ws client.
 * @property callback            The inner callback.
 */
class WebsocketInterceptorCallback<T>(
	private var websocketClient: WebsocketClient,
	var callback: WebsocketCallback<T>
) : WebsocketCallback<T> {
	/** Event to handle connection failure and try to reconnect. */
	var connectionHandler: WebsocketEventHandler = WebsocketConnectionHandler(websocketClient, this)
	
	/** Event to handle no response from server. */
	private var noResponseHandler: WebsocketEventHandler = WebsocketNoResponseHandler(websocketClient, this)
	
	/** Event to handle failure in the disconnection state. */
	private var disconnectionHandler: WebsocketEventHandler = WebsocketDisconnectionHandler(websocketClient, this)
	
	/** Forces the call of onClosing and onClosed. */
	lateinit var forceClosingHandler: WebsocketEventHandler
	
	/** Tells the interceptor if the closing has been made by the client. */
	var closedByClient: Boolean = false
	
	/** Has the onClosing handler been called. */
	var onClosingCalled: Boolean = false
	
	/** Socket using the callback. */
	internal var socket: WebSocket? = null
	
	override fun onMessage(message: T) {
		noResponseHandler.run()
		callback.onMessage(message)
	}
	
	override fun onClosing(websocketCloseObject: WebsocketCloseObject) {
		forceClosingHandler.cancel()
		onClosingCalled = true
		// we force disconnection
		disconnectionHandler.run()
		callback.onClosing(websocketCloseObject)
	}
	
	override fun onClosed(websocketCloseObject: WebsocketCloseObject) {
		// we stop disconnection forcing
		disconnectionHandler.cancel()
		callback.onClosed(websocketCloseObject) // Reconnect if server closed stream and keepAlive is on
		if (websocketClient.configuration.isKeepAlive && !closedByClient) {
			websocketClient.open()
		}
		socket = null
	}
	
	override fun onOpen(response: Response) {
		connectionHandler.cancel()
		noResponseHandler.run()
		callback.onOpen(response)
	}
	
	override fun onFailure(exception: ApiException) = callback.onFailure(exception)
}